File: study/ts/script/ui/GroupsTipBar.ts
import { G_SceneManager, G_ServerTime, G_TopLevelNode } from "../init";
import { Lang } from "../lang/Lang";
import ViewBase from "../scene/ViewBase";
import { handler } from "../utils/handler";
import UIHelper from "../utils/UIHelper";
import CommonButton from "./component/CommonButton";
import CommonButtonCountDown from "./component/CommonButtonCountDown";
import CommonHeadFrame from "./component/CommonHeadFrame";
import CommonHeroIcon from "./component/CommonHeroIcon";

const { ccclass, property } = cc._decorator;
@ccclass
export default class GroupsTipBar extends ViewBase {

    @property({ type: cc.Node, visible: true })
    _panelRoot: cc.Node = null;

    @property({ type: cc.Sprite, visible: true })
    _imageBg: cc.Sprite = null;

    @property({ type: cc.Sprite, visible: true })
    _imageInvite: cc.Sprite = null;

    @property({ type: CommonHeroIcon, visible: true })
    _icon: CommonHeroIcon = null;

    @property({ type: CommonHeadFrame, visible: true })
    _headFrame: CommonHeadFrame = null;

    @property({ type: cc.Label, visible: true })
    _leaderName: cc.Label = null;

    @property({ type: CommonButtonCountDown, visible: true })
    _btnCancel: CommonButtonCountDown = null;

    @property({ type: CommonButton, visible: true })
    _btnOk: CommonButton = null;

    @property({ type: cc.Label, visible: true })
    _targetName: cc.Label = null;

    @property({ type: cc.Sprite, visible: true })
    _imageTips: cc.Sprite = null;

    @property({ type: cc.Toggle, visible: true })
    _checkBoxTip: cc.Toggle = null;

    _rootY: number;
    _rootX: any;
    onCreate() {
        this.setSceneSize();
        this._rootY = 0;
        this._btnCancel.setString(Lang.get('groups_refuse'));
        this._btnOk.setString(Lang.get('groups_accept'));
        var sceneSize = this.getSceneSize();
        var noteSize = this._panelRoot.getContentSize();
        this._rootX = this._panelRoot.x;
        // this._rootY = sceneSize.height - noteSize.height - 66;
        this._rootY =  66;
        this._panelRoot.y = (this._rootY);
    }
    onEnter() {
    }
    onExit() {
    }
    _checkFilter(filterNames) {
        filterNames = filterNames || {};
        var runningScene = G_SceneManager.getRunningScene();
        var sceneName = runningScene.getName();
        for (var i in filterNames) {
            var filterName = filterNames[i];
            if (sceneName == filterName) {
                return true;
            }
        }
        return false;
    }
    slideOut(data, filterViewNames) {
        if (this._checkFilter(filterViewNames)) {
            return;
        }
        G_TopLevelNode.addToGroupNoticeLayer(this.node);
        this._updateUI(data);
        this._moveOut();
    }
    _updateUI(data) {
        var show = data.showImageTips || false;
        this._imageTips.node.active = (show);
        if (data.imageRes) {
            UIHelper.loadTexture(this._imageInvite, data.imageRes);
        }
        if (data.name) {
            this._leaderName.string = (data.name);
        }
        if (data.nameColor) {
            this._leaderName.node.color = (data.nameColor);
        }
        if (data.targetName) {
            this._targetName.string = (data.targetName);
        }
        if (data.covertId) {
            this._icon.updateUI(data.covertId, null, data.limitLevel);
        }
        if (data.headFrameId) {
            this._headFrame.updateUI(data.headFrameId, this._icon.node.scale);
        }
        if (data.level) {
            this._headFrame.setLevel(data.level);
        }
        if (data.endTime) {
            this._btnCancel.startCountDown(data.endTime, handler(this, this._countDownEnd), handler(this, this._countDownFormatStr));
        }
        if (data.imageBg) {
            UIHelper.loadTexture(this._imageBg, data.imageBg);
        }
    }
    _countDownEnd() {
        this.closeWindow();
    }
    _countDownFormatStr(endTime) {
        var time = G_ServerTime.getLeftSeconds(endTime);
        var str = '';
        if (time < 10) {
            str = ' ';
        }
        str = str + (time + 's');
        return str;
    }
    _moveOut() {
        var sceneSize = this.getSceneSize();
        var noteSize = this._panelRoot.getContentSize();
        var posX = sceneSize.width/2 - noteSize.width;
        var posY = this._rootY;
        var callAction = cc.callFunc(function () {
        });
        var action = cc.moveTo(0.3, cc.v2(posX, posY));
        var runningAction = cc.sequence(action, callAction);
        this._panelRoot.runAction(runningAction);
    }
    closeWindow() {
        var sceneSize = this.getSceneSize();
        var posX = sceneSize.width/2 + 10;
        var callAction = cc.callFunc(function () {
            this.node.destroy();
        }.bind(this));
        var action = cc.moveBy(0.3, cc.v2(posX, 0));
        var runningAction = cc.sequence(action, callAction);
        this._panelRoot.runAction(runningAction);
    }
    onBtnCancel() {
    }
    onBtnOk() {
    }
    onCheckBoxClicked() {
    }
}

File: study/lua/src/app/ui/GroupsTipBar.lua

-- Author: liangxu
-- Date:2019-3-7
-- Describle：

local ViewBase = require("app.ui.ViewBase")
local GroupsTipBar = class("GroupsTipBar", ViewBase)

function GroupsTipBar:ctor()
	local resource = {
		file = Path.getCSB("GroupsMainNode", "groups"),
		binding = {
			_btnCancel = {
				events = {{event = "touch", method = "onBtnCancel"}}
			},
			_btnOk = {
				events = {{event = "touch", method = "onBtnOk"}}
			},
		},
	}
	GroupsTipBar.super.ctor(self, resource)
end

function GroupsTipBar:onCreate()
	self._rootY = 0

	self._btnCancel:setString(Lang.get("groups_refuse"))
	self._btnOk:setString(Lang.get("groups_accept"))
	self._checkBoxTip:addEventListener(handler(self, self.onCheckBoxClicked))

	local sceneSize = self:getSceneSize()
	local noteSize = self._panelRoot:getContentSize()
	self._rootX = self._panelRoot:getPositionX()
	self._rootY = sceneSize.height - noteSize.height - 66
	self._panelRoot:setPositionY(self._rootY)
end

function GroupsTipBar:onEnter()
	
end

function GroupsTipBar:onExit()

end

function GroupsTipBar:_checkFilter(filterNames)
	filterNames = filterNames or {}
	local runningScene = G_SceneManager:getRunningScene()
	local sceneName = runningScene:getName()
	for i, filterName in ipairs(filterNames) do
		if sceneName == filterName then
			return true
		end
	end
	return false
end

function GroupsTipBar:slideOut(data, filterViewNames)
	if self:_checkFilter(filterViewNames) then
		return
	end
	G_TopLevelNode:addToGroupNoticeLayer(self)
	self:_updateUI(data)
	self:_moveOut()
end

function GroupsTipBar:_updateUI(data)
	local show = data.showImageTips or false
	self._imageTips:setVisible(show)
	if data.imageRes then
		self._imageInvite:loadTexture(data.imageRes)
	end
	if data.name then
		self._leaderName:setString(data.name)
	end
	if data.nameColor then
		self._leaderName:setColor(data.nameColor)
	end
	if data.targetName then
		self._targetName:setString(data.targetName)
	end
	if data.covertId then
		self._icon:updateUI(data.covertId, nil, data.limitLevel, data.limitRedLevel)
	end
	if data.headFrameId then
		self._headFrame:updateUI(data.headFrameId,self._icon:getScale())
	end
	if data.level then
		self._headFrame:setLevel(data.level)
	end
	if data.endTime then
		self._btnCancel:startCountDown(data.endTime, handler(self, self._countDownEnd), handler(self, self._countDownFormatStr))
	end
	if data.imageBg then
		self._imageBg:loadTexture(data.imageBg)
	end

end

function GroupsTipBar:_countDownEnd()
	self:closeWindow()
end

function GroupsTipBar:_countDownFormatStr(endTime)
	local time = G_ServerTime:getLeftSeconds(endTime)
	local str = ""
	if time < 10 then 
		str = " " 
	end
	str = str .. time .. "s"
	return str
end

function GroupsTipBar:_moveOut()
	local sceneSize = self:getSceneSize()
	local noteSize = self._panelRoot:getContentSize()
	local posX = sceneSize.width - noteSize.width
	local posY = self._rootY
	local callAction = cc.CallFunc:create(function()

	end)
	local action = cc.MoveTo:create(0.3,cc.p(posX,posY))
	local runningAction = cc.Sequence:create(action,callAction)
	self._panelRoot:runAction(runningAction)
end

function GroupsTipBar:closeWindow()
	local sceneSize = self:getSceneSize()
	local posX = sceneSize.width + 10
	local callAction = cc.CallFunc:create(function()
		self:removeFromParent()
	end)
	local action = cc.MoveBy:create(0.3,cc.p(posX,0))
	local runningAction = cc.Sequence:create(action,callAction)
	self._panelRoot:runAction(runningAction)
end

function GroupsTipBar:onBtnCancel()
	
end

function GroupsTipBar:onBtnOk()
	
end

function GroupsTipBar:onCheckBoxClicked()
	
end

return GroupsTipBar